Intermittent failure of cpp test jobs in CI

### 👓 What did you see?

Observed instances where the cpp jobs appear to have failed - with changes being unrelated to the cpp job itself e.g. modifying markdown files. For example, the following [failed run](https://github.com/cucumber/gherkin/actions/runs/7545566347/job/20541479511?pr=215) on #215. [Another failed run](https://github.com/cucumber/gherkin/actions/runs/7544434482/job/20537758755). Re-running the job for other PRs resulted in the job succeeding.

Download of latest Ninja build failing. Perhaps rate limiting or network issue?

```console
Run cmake -P cmake/cmate --cc=clang install
CMate: using compilers CC=clang CXX=clang++
CMate: checking nlohmann/json
Cloning into '/Users/runner/work/gherkin/gherkin/cpp/.cenv/downloads/nlohmann_json/sources'...
CMate: building with: -DJSON_BuildTests=OFF
CMake Error at cmake/cmate:23 (message):
  CMate: error: download of
  https://api.github.com/repos/ninja-build/ninja/releases/latest failed:
  22;"HTTP response code said error"
Call Stack (most recent call first):
  cmake/cmate:186 (cmate_die)
  cmake/cmate:212 (cmate_download)
  cmake/cmate:262 (cmate_github_get_latest)
  cmake/cmate:282 (cmate_check_ninja)
  cmake/cmate:1127 (cmate_set_ninja)
  cmake/cmate:1256 (cmate_install_cmake_dep)
  cmake/cmate:1270 (cmate_install_dep)
  cmake/cmate:1315 (cmate_install_repo)
  cmake/cmate:1474 (cmate_install)
  cmake/cmate:1474 (cmake_language)
  cmake/cmate:1493 (cmate_process_cmd)


Error: Process completed with exit code 1.
```

### ✅ What did you expect to see?

Cpp job succeeds.

### 📦 Which tool/library version are you using?

Running with tooling versioned in repo against commit 579c66d.

### 🔬 How could we reproduce it?

Intermittent failure with pipelines. Not sure of root cause. Issue should arise with subsequent pipelines.

### 📚 Any additional context?

Apologies if raised under an existing issue - have not yet located.

--- a/CHANGELOG.md
+++ b/CHANGELOG.md
@@ -10,6 +10,7 @@ This document is formatted according to the principles of [Keep A CHANGELOG](htt
 ### Fixed
 - Provide trailing space in Irish keywords ([#243](https://github.com/cucumber/gherkin/pull/243))
 - Tamil "And" and "But" translations should have single trailing space ([#243](https://github.com/cucumber/gherkin/pull/243))
+- Intermittent failure of cpp test jobs in CI ([#217](https://github.com/cucumber/gherkin/issues/217))
 
 ## [28.0.0] - 2024-02-15
 ### Added

--- a/cpp/CMakeLists.txt
+++ b/cpp/CMakeLists.txt
@@ -9,6 +9,10 @@ set(CMAKE_CXX_STANDARD_REQUIRED ON)
 set(CMAKE_CXX_EXTENSIONS OFF)
 set(CMAKE_POSITION_INDEPENDENT_CODE ON)
 
+if (CMAKE_CXX_COMPILER_ID STREQUAL "MSVC")
+    add_compile_definitions(_CRT_SECURE_NO_WARNINGS _SCL_SECURE_NO_WARNINGS)
+endif()
+
 find_package(nlohmann_json CONFIG REQUIRED)
 find_package(cucumber_messages CONFIG REQUIRED)
 

--- a/cpp/cmake/cmate
+++ b/cpp/cmake/cmate
@@ -47,6 +47,10 @@ function(cmate_setgdir VAR VAL)
     file(MAKE_DIRECTORY ${${VAR}})
 endfunction()
 
+function(cmate_sleep DURATION)
+    execute_process(COMMAND ${CMAKE_COMMAND} -E sleep ${DURATION})
+endfunction()
+
 function(cmate_load_version)
     if(NOT "${CMATE_VERSION}" STREQUAL "")
         return()
@@ -176,15 +180,37 @@ endfunction()
 function(cmate_download URL FILE)
     if(CMATE_SIMULATE)
         cmate_msg("download ${URL} to ${FILE}")
-    else()
+        return()
+    endif()
+
+    set(WAIT_INTERVAL 5)
+    set(MAX_RETRIES 10)
+    set(RETRIES ${MAX_RETRIES})
+
+    cmate_msg("downloading ${URL}")
+
+    while(1)
         file(DOWNLOAD ${URL} ${FILE} STATUS ST)
-    endif()
 
-    list(GET ST 0 RC)
+        list(GET ST 0 RC)
 
-    if(RC)
-        cmate_die("download of ${URL} failed: ${ST}")
-    endif()
+        if(RC)
+            if(RETRIES GREATER 1)
+                math(EXPR RETRIES "${RETRIES} - 1")
+                math(EXPR ATTEMPT "${MAX_RETRIES} - ${RETRIES}")
+                cmate_msg(
+                    "download of ${URL} failed"
+                    " (attempt ${ATTEMPT} of ${MAX_RETRIES}"
+                    ", retrying in ${WAIT_INTERVAL}s)"
+                )
+                cmate_sleep(${WAIT_INTERVAL})
+            else()
+                cmate_die("download of ${URL} failed: ${ST}")
+            endif()
+        else()
+            break()
+        endif()
+    endwhile()
 endfunction()
 
 function(cmate_set_build_type RELEASE_FLAG_VAR)
@@ -202,41 +228,16 @@ function(cmate_set_build_type RELEASE_FLAG_VAR)
     cmate_setg(CMATE_BUILD_DIR "${CMATE_BUILD_BASE_DIR}/${TDIR}")
 endfunction()
 
-function(cmate_github_get_latest REPO VAR RE)
-    set(URL "https://api.github.com/repos/${REPO}/releases/latest")
-    set(TDIR "${CMATE_TMP_DIR}/${REPO}")
-    set(INFO "${TDIR}/info.json")
+function(cmate_github_get_latest REPO PKG VAR)
+    set(URL "https://github.com/${REPO}/releases/latest/download/${PKG}")
 
-    if (NOT EXISTS ${INFO})
-        file(MAKE_DIRECTORY ${TDIR})
-        cmate_download(${URL} ${INFO})
+    set(FILE "${CMATE_DL_DIR}/${PKG}")
+
+    if (NOT EXISTS ${FILE})
+        cmate_download(${URL} ${FILE})
     endif()
 
-    file(READ ${INFO} VINFO)
-    cmate_json_get_array(${VINFO} "assets" ASSETS)
-
-    foreach(ASSET ${ASSETS})
-        string(
-            JSON
-            BDURL
-            ERROR_VARIABLE ERR
-            GET "${ASSET}" "browser_download_url"
-        )
-
-        if(NOT ERR AND ${BDURL} MATCHES ${RE})
-            string(JSON FILE GET "${ASSET}" "name")
-            set(FILE "${CMATE_DL_DIR}/${FILE}")
-
-            if (NOT EXISTS ${FILE})
-                cmate_download(${BDURL} ${FILE})
-            endif()
-
-            set(${VAR} ${FILE} PARENT_SCOPE)
-            break()
-        endif()
-    endforeach()
-
-    file(REMOVE_RECURSE ${TDIR})
+    set(${VAR} ${FILE} PARENT_SCOPE)
 endfunction()
 
 function(cmate_check_ninja VAR)
@@ -259,11 +260,7 @@ function(cmate_check_ninja VAR)
         endif()
 
         if(NOT EXISTS "${CMATE_ENV_BIN_DIR}/${NCMD}")
-            cmate_github_get_latest(
-                "ninja-build/ninja"
-                NZIP
-                "ninja-${NOS}.zip$"
-            )
+            cmate_github_get_latest("ninja-build/ninja" "ninja-${NOS}.zip" NZIP)
 
             file(REMOVE_RECURSE ${TDIR})
             file(ARCHIVE_EXTRACT INPUT ${NZIP} DESTINATION ${TDIR})

